File: app/src/main/java/org/adaway/db/dao/HostsSourceDao.java
package org.adaway.db.dao;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.Query;
import androidx.room.Update;

import org.adaway.db.entity.HostsSource;

import java.time.ZonedDateTime;
import java.util.List;
import java.util.Optional;

import static androidx.room.OnConflictStrategy.IGNORE;

/**
 * This interface is the DAO for {@link HostsSource} entities.
 *
 * @author Bruce BUJON (bruce.bujon(at)gmail(dot)com)
 */
@Dao
public interface HostsSourceDao {
    @Insert(onConflict = IGNORE)
    void insert(HostsSource source);

    @Update
    void update(HostsSource source);

    @Delete
    void delete(HostsSource source);

    @Query("SELECT * FROM hosts_sources WHERE enabled = 1 AND id != 1 ORDER BY url ASC")
    List<HostsSource> getEnabled();

    default void toggleEnabled(HostsSource source) {
        int id = source.getId();
        boolean enabled = !source.isEnabled();
        source.setEnabled(enabled);
        setSourceEnabled(id, enabled);
        setSourceItemsEnabled(id, enabled);
    }

    @Query("UPDATE hosts_sources SET enabled = :enabled WHERE id =:id")
    void setSourceEnabled(int id, boolean enabled);

    @Query("UPDATE hosts_lists SET enabled = :enabled WHERE source_id =:id")
    void setSourceItemsEnabled(int id, boolean enabled);

    @Query("SELECT * FROM hosts_sources WHERE id = :id")
    Optional<HostsSource> getById(int id);

    @Query("SELECT * FROM hosts_sources WHERE id != 1 ORDER BY label ASC")
    List<HostsSource> getAll();

    @Query("SELECT * FROM hosts_sources WHERE id != 1 ORDER BY label ASC")
    LiveData<List<HostsSource>> loadAll();

    @Query("UPDATE hosts_sources SET last_modified_online = :dateTime WHERE id = :id")
    void updateOnlineModificationDate(int id, ZonedDateTime dateTime);

    @Query("UPDATE hosts_sources SET last_modified_local = :localModificationDate, last_modified_online = :onlineModificationDate WHERE id = :id")
    void updateModificationDates(int id, ZonedDateTime localModificationDate, ZonedDateTime onlineModificationDate);

    @Query("UPDATE hosts_sources SET size = (SELECT count(id) FROM hosts_lists WHERE source_id = :id) WHERE id = :id")
    void updateSize(int id);

    @Query("SELECT count(id) FROM hosts_sources WHERE enabled = 1 AND last_modified_online > last_modified_local")
    LiveData<Integer> countOutdated();

    @Query("SELECT count(id) FROM hosts_sources WHERE enabled = 1 AND last_modified_online <= last_modified_local")
    LiveData<Integer> countUpToDate();

    @Query("UPDATE hosts_sources SET last_modified_local = null, size = 0 WHERE id = :id")
    void clearProperties(int id);
}

File: app/src/main/java/org/adaway/db/dao/HostsSourceDao.kt
package org.adaway.db.dao

import androidx.lifecycle.LiveData
import androidx.room.Dao
import androidx.room.Delete
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query
import androidx.room.Update
import org.adaway.db.entity.HostsSource
import java.time.ZonedDateTime
import java.util.Optional

@Dao
interface HostsSourceDao {
    @Insert(onConflict = OnConflictStrategy.IGNORE)
    fun insert(source: HostsSource)

    @Update
    fun update(source: HostsSource)

    @Delete
    fun delete(source: HostsSource)

    @get:Query("SELECT * FROM hosts_sources WHERE enabled = 1 AND id != 1 ORDER BY url ASC")
    val enabled: List<HostsSource>

    fun toggleEnabled(source: HostsSource) {
        val id = source.id
        val enabled = !source.isEnabled
        source.isEnabled = enabled
        setSourceEnabled(id, enabled)
        setSourceItemsEnabled(id, enabled)
    }

    @Query("UPDATE hosts_sources SET enabled = :enabled WHERE id =:id")
    fun setSourceEnabled(id: Int, enabled: Boolean)

    @Query("UPDATE hosts_lists SET enabled = :enabled WHERE source_id =:id")
    fun setSourceItemsEnabled(id: Int, enabled: Boolean)

    @Query("SELECT * FROM hosts_sources WHERE id = :id")
    fun getById(id: Int): Optional<HostsSource>

    @get:Query("SELECT * FROM hosts_sources WHERE id != 1 ORDER BY label ASC")
    val all: List<HostsSource>

    @Query("SELECT * FROM hosts_sources WHERE id != 1 ORDER BY label ASC")
    fun loadAll(): LiveData<List<HostsSource>>

    @Query("UPDATE hosts_sources SET last_modified_online = :dateTime WHERE id = :id")
    fun updateOnlineModificationDate(id: Int, dateTime: ZonedDateTime)

    @Query("UPDATE hosts_sources SET last_modified_local = :localModificationDate, last_modified_online = :onlineModificationDate WHERE id = :id")
    fun updateModificationDates(
        id: Int,
        localModificationDate: ZonedDateTime,
        onlineModificationDate: ZonedDateTime
    )

    @Query("UPDATE hosts_sources SET entityTag = :entityTag WHERE id = :id")
    fun updateEntityTag(id: Int, entityTag: String)

    @Query("UPDATE hosts_sources SET size = (SELECT count(id) FROM hosts_lists WHERE source_id = :id) WHERE id = :id")
    fun updateSize(id: Int)

    @Query("SELECT count(id) FROM hosts_sources WHERE enabled = 1 AND last_modified_online > last_modified_local")
    fun countOutdated(): LiveData<Int>

    @Query("SELECT count(id) FROM hosts_sources WHERE enabled = 1 AND last_modified_online <= last_modified_local")
    fun countUpToDate(): LiveData<Int>

    @Query("UPDATE hosts_sources SET last_modified_local = NULL, last_modified_online = NULL, entityTag = NULL, size = 0 WHERE id = :id")
    fun clearProperties(id: Int)
}
